pg_search/src/postgres/transaction.rs

```diff
@@ -137,18 +137,57 @@ unsafe extern "C" fn pg_search_xact_callback(
 }
 
 fn finalize_drop<W: WriterClient<WriterRequest> + Send + Sync + 'static>(
-    writer: &Arc<Mutex<W>>,
+    _writer: &Arc<Mutex<W>>,
     directory: &WriterDirectory,
 ) -> Result<(), ClientError> {
-    // ask the remote side to drop the index
-    writer.lock().request(WriterRequest::DropIndex {
-        directory: directory.clone(),
-    })?;
+    let writer = unsafe { SearchIndex::get_writer() };
 
-    // then tell it to do it now
-    writer.lock().request(WriterRequest::Commit {
-        directory: directory.clone(),
-    })?;
+    writer
+        .drop_index(directory.clone())
+        .expect("must be able to finalize drop");
+    writer
+        .commit(directory.clone())
+        .expect("commit must work in finalize drop");
 
     Ok(())
 }
+
+/// Create a new index writer advisory lock.
+/// Advisory locks are global across the entire Postgres cluster,
+/// so its important that the key is qualified with the database oid.
+pub struct IndexWriterLock {
+    database_oid: u32,
+    index_oid: u32,
+}
+
+impl IndexWriterLock {
+    pub fn new(database_oid: u32, index_oid: u32) -> Self {
+        Self {
+            database_oid,
+            index_oid,
+        }
+    }
+
+    // This method combines the two OIDs into a single 64-bit lock key
+    fn key(&self) -> i64 {
+        // Shift the database_oid into the high 32 bits, then combine with index_oid
+        ((self.database_oid as i64) << 32) | (self.index_oid as i64)
```

Comment:
👍 this is good.

Just as a thing to think about... using advisory locks like this kinda means pg_search is now the only extension allowed to use advisory locks within a database.  These lock ids aren't extension-specific -- they're database wide.  Something for us to keep in mind.  I have no idea how many people in the world actually use advisory locks in their own database applications.  I have used them before in that kind of scenario, but I doubt I'm representative of the world at large.